virtual_ecosystem/models/animal/animal_model.py

```diff
@@ -769,131 +774,231 @@ def remove_dead_cohort_community(self) -> None:
 
         Args:
-            parent_cohort: The AnimalCohort instance which is producing a new cohort.
+            parent_cohort: The parent cohort giving birth.
         """
-        # Handle semelparous reproduction (where parents die after reproduction)
-        non_reproductive_mass_loss_c = non_reproductive_mass_loss_n = (
-            non_reproductive_mass_loss_p
-        ) = 0.0
-
-        if parent_cohort.functional_group.reproductive_type == "semelparous":
-            loss_c = (
-                parent_cohort.mass_cnp.carbon
-                * parent_cohort.constants.semelparity_mass_loss
-            )
-            loss_n = (
-                parent_cohort.mass_cnp.nitrogen
-                * parent_cohort.constants.semelparity_mass_loss
-            )
-            loss_p = (
-                parent_cohort.mass_cnp.phosphorus
-                * parent_cohort.constants.semelparity_mass_loss
-            )
+        reproductive_mass = self.calculate_total_reproductive_mass(parent_cohort)
+        number_offspring = self.calculate_offspring_count(
+            parent_cohort, reproductive_mass
+        )
 
-            # Cap loss to available mass
-            non_reproductive_mass_loss_c = min(loss_c, parent_cohort.mass_cnp.carbon)
-            non_reproductive_mass_loss_n = min(loss_n, parent_cohort.mass_cnp.nitrogen)
-            non_reproductive_mass_loss_p = min(
-                loss_p, parent_cohort.mass_cnp.phosphorus
-            )
+        if number_offspring <= 0:
+            return  # Insufficient mass for offspring
 
-            # Reduce parent's mass in-place
-            parent_cohort.mass_cnp.update(
-                carbon=-non_reproductive_mass_loss_c,
-                nitrogen=-non_reproductive_mass_loss_n,
-                phosphorus=-non_reproductive_mass_loss_p,
-            )
+        self.handle_offspring_creation(parent_cohort, number_offspring)
+        self.handle_post_birth_parent_updates(parent_cohort, number_offspring)
 
-            # Kill semelparous parent cohort
-            parent_cohort.is_alive = False
+    def calculate_total_reproductive_mass(
+        self, parent: AnimalCohort
+    ) -> dict[str, float]:
+        """Calculate total reproductive mass available for offspring.
 
-        # Calculate total available reproductive mass
-        total_reproductive_mass_c = (
-            parent_cohort.reproductive_mass_cnp.carbon + non_reproductive_mass_loss_c
-        )
-        total_reproductive_mass_n = (
-            parent_cohort.reproductive_mass_cnp.nitrogen + non_reproductive_mass_loss_n
+        For semelparous species, part of the parent's non-reproductive mass
+        is also transferred to reproduction as they die after reproducing.
+
+        Args:
+            parent: The parent cohort.
+
+        Returns:
+            Reproductive mass for carbon, nitrogen, phosphorus (kg).
+        """
+        semelparous_loss = self.calculate_semelparous_mass_loss(parent)
+
+        return {
```

Comment:
I seem to remember a carbon-nitrogen-phosphorus dataclass being defined somewhere else (maybe for the soil model?). Maybe we want to reuse that, for consistency (although that might be done in a separate PR).

Reply:
We've been looking at the same thing in `plants` - we do though have the issue that we _might_ want to run a model without animals (although animals without plants is a bit less feasible 🤣 ) so sharing structures between models is problematic?

Actually it isn't is it -  we don't have to be using the model to import from it. It is only a problem if somebody plugs in a new model that _replaces_ the existing one and doesn't provide the object. And that's clearly somebody else's problem at that point.

I think there might be a case for some `core` structures like this, but that is definitely a different PR!